File: java/uastack.java
package uas.java;

// iki dicopy ngisor e pakage
import java.util.Scanner;

/**
 * 1. hapus kabeh seng public class seng pakage karo autor ditokne
 * 
 */

 // iki kabeh dicopy
class Stack {
    int top;
    String[] data;

    public Stack() {
        top = -1;
        data = new String[10];
    }
}

// uastack iki dihanti karo jeneng file mu contone BinaryTree kemau gede cilik e podo
public class uastack { 
    static Stack piringBersih;
    static Stack piringKotor;

    public static void init() {
        piringBersih = new Stack();
        piringKotor = new Stack();
    }

    public static boolean isEmpty(Stack stack) {
        return stack.top == -1;
    }

    public static boolean isFull(Stack stack) {
        return stack.top == 9;
    }

    public static void DataPiringBersih() {
        for (int i = 0; i < 5; i++) {
            piringBersih.data[i] = "Piring Bersih " + (i + 1);
            piringBersih.top++;
        }
    }

    public static void Push(Stack stack) {
        if (!isFull(stack)) {
            Scanner scanner = new Scanner(System.in);
            System.out.print("Masukkan Piring : ");
            String piring = scanner.nextLine();
            stack.data[++stack.top] = piring;
            System.out.println();
        } else {
            System.out.println("Tumpukan Penuh!");
        }
    }

    public static void Pop(Stack stack1, Stack stack2) {
        if (!isEmpty(stack1)) {
            String topPiring = stack1.data[stack1.top];
            stack1.top--;
            if (!isFull(stack2)) {
                stack2.data[++stack2.top] = topPiring;
                System.out.println("Piring " + topPiring + " dipindahkan ke tumpukan piring kotor.");
            } else {
                System.out.println("Tumpukan piring kotor penuh!");
            }
        } else {
            System.out.println("Tumpukan piring bersih kosong!");
        }
    }

    public static void Clear(Stack stack) {
        if (!isEmpty(stack)) {
            stack.top = -1;
            System.out.println("Tumpukan telah dikosongkan!");
        } else {
            System.out.println("Tumpukan sudah kosong!");
        }
    }


    public static void Display(Stack stack, String namaStack) {
        if (!isEmpty(stack)) {
            System.out.println("Isi " + namaStack + ": ");
            for (int i = 0; i <= stack.top; i++) {
                System.out.println(stack.data[i]);
            }
        } else {
            System.out.println(namaStack + " Kosong");
        }

        if (isFull(stack)) {
            System.out.println(namaStack + " Penuh");
        }
    }


    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int pilih;
        init();
        DataPiringBersih();

        do {
            System.out.println("Stock Piring Bersih Dan Kotor");
            Display(piringBersih, "Piring Bersih");
            System.out.println();
            Display(piringKotor, "Piring Kotor");
            System.out.println();
            System.out.println("Menu :");
            System.out.println("1. Masukkan piring bersih");
            System.out.println("2. Masukkan piring kotor");
            System.out.println("3. Ambil piring bersih");
            System.out.println("4. Ambil piring kotor");
            System.out.println("5. Clear tumpukan piring bersih");
            System.out.println("6. Clear tumpukan piring kotor");
            System.out.println("7. Keluar");
            System.out.print("Pilihan:");
            pilih = scanner.nextInt();

            switch (pilih) {
                case 1:
                    Push(piringBersih);
                    break;
                case 2:
                    Push(piringKotor);
                    break;
                case 3:
                    Pop(piringBersih, piringKotor);
                    break;
                case 4:
                    Pop(piringKotor, piringBersih);
                    break;
                case 5:
                    Clear(piringBersih);
                    break;
                case 6:
                    Clear(piringKotor);
                    break;
                case 7:
                    System.out.println("Terima kasih!");
                    return;
                default:
                    System.out.println("Pilihan tidak ada!");
                    break;
            }
        } while (true);
    }
}

File: C++/uastack.cpp
#include <iostream>
#include <string>
#define MAX_STACK 10

using namespace std;

struct Stack {
    int top;
    string data[MAX_STACK];
};

Stack piringBersih;
Stack piringKotor;

void init() {
    piringBersih.top = -1;
    piringKotor.top = -1;
}

int isEmpty(Stack stack) {
    return stack.top == -1;
}

int isFull(Stack stack) {
    return stack.top == MAX_STACK - 1;
}

void DataPiringBersih() {
    for (int i = 0; i < 5; i++) {
        piringBersih.data[i] = "Piring Bersih " + to_string(i + 1);
        piringBersih.top++;
    }
}


void push(Stack& stack) {
    if (!isFull(stack)) {
        string piring;
        cout << "Masukkan Piring: ";
        cin.ignore();
        getline(cin, piring);
        stack.data[++stack.top] = piring;
        cout << endl;
    }
    else {
        cout << "Tumpukan penuh!" << endl;
    }
}

void pop(Stack& stack1, Stack& stack2) {
    if (!isEmpty(stack1)) {
        string toppiring = stack1.data[stack1.top];
        stack1.top--;
        if (!isFull(stack2)) {
            stack2.data[++stack2.top] = toppiring;
            cout << "Piring " << toppiring << " dipindahkan ke tumpukan piring kotor." << endl;
        }
        else {
            cout << "Tumpukan piring kotor penuh!" << endl;
        }
    }
    else {
        cout << "Tumpukan piring bersih kosong!" << endl;
    }
}

void clear(Stack& stack) {
    if (!isEmpty(stack)) {
        stack.top = -1;
        cout << "Tumpukan telah dikosongkan" << endl;
    }
    else {
        cout << "Tumpukan sudah kosong" << endl;
    }
}

void display(Stack stack, string namaStack) {
    if (!isEmpty(stack)) {
        cout << "Isi " << namaStack << ": " << endl;
        for (int i = 0; i <= stack.top; i++) {
            cout << stack.data[i] << endl;
        }
    }
    else {
        cout << namaStack << " kosong" << endl;
    }

    if (isFull(stack)) {
        cout << namaStack << " penuh" << endl;
    }
}


int main() {
    int pilih;
    init();
    DataPiringBersih();

    do {
        system("cls");
        cout << "Stock Piring Bersih Dan Kotor" << endl;
        display(piringBersih, "Piring Bersih");
        cout << endl;
        display(piringKotor, "Piring Kotor");
        cout << endl;
        cout << "Menu: " << endl;
        cout << "1. Masukkan piring bersih" << endl;
        cout << "2. Masukkan piring kotor" << endl;
        cout << "3. Ambil piring bersih" << endl;
        cout << "4. Ambil piring kotor" << endl;
        cout << "5. Clear tumpukan piring bersih" << endl;
        cout << "6. Clear tumpukan piring kotor" << endl;
        cout << "7. Keluar" << endl;
        cout << "Pilihan: ";
        cin >> pilih;

        switch (pilih) {
        case 1:
            push(piringBersih);
            break;
        case 2:
            push(piringKotor);
            break;
        case 3:
            pop(piringBersih, piringKotor);
            break;
        case 4:
            pop(piringKotor, piringBersih);
            break;
        case 5:
            clear(piringBersih);
            break;
        case 6:
            clear(piringKotor);
            break;
        case 7:
            cout << "Terima kasih!" << endl;
            return 0;
        default:
            cout << "Pilihan tidak ada!" << endl;
            break;
        }
    } while (true);
}
